### ISSUE TITLE
Custom Scalar as argument

### ISSUE THREAD

**Issue by ashleyw** [19majkel94/type-graphql#65]
Is it possible to use a custom scalar as an argument in a resolver? It appears like `convertToType` expects a `ClassType`, but `GraphQLScalarType` returns an object?

Thanks!

**Comment by ashleyw** [comment]
Ah! I needed to map the scalar in `buildSchema`'s `scalarsMap` property, and specify the underlying class as the `@Arg` type, rather than the `GraphQLScalarType`

**Comment by 19majkel94** [comment]
Yes, you can provide the custom scalar to the `@Arg` decorator or map it with the class if custom scalar convert from/to class (like Date or ObjectId).
```ts
class SampleResolver {
  @Query()
  sampleQuery(@Arg(type => CustomScalar) data: any): boolean {
    return true;
  }
}
```

**Comment by ashleyw** [comment]
@19majkel94 are you sure the former is true? In the `convertToType` method it tries to call `new CustomScalar()`, where `CustomScalar` isn't a function/class but instead an object provided by `GraphQLScalarType`.

**Comment by 19majkel94** [comment]
For getting query/mutation/resolver arg, it calls `getGraphQLInputType`, which calls `convertTypeIfScalar`, which checks:
```ts
if (type instanceof GraphQLScalarType) {
  return type;
}
```

**Comment by ashleyw** [comment]
That's strange. I've just experienced this same issue again, check this out:


**Custom Scalar:**

```ts
import { GraphQLScalarType, Kind } from 'graphql';
import { PaginationDirection } from '../input/PaginationArgs';

export interface CursorID {
  id: number | string;
  direction: PaginationDirection;
}

export interface CursorPaged {
  skip: number;
}

export type Cursor = CursorID | CursorPaged;

export const CursorScalar = new GraphQLScalarType({
  name: 'Cursor',
  description: 'Pagination cursor',

  parseValue(value: string): Cursor {
    return JSON.parse(Buffer.from(value, 'base64').toString('ascii'));
  },

  serialize(value: Cursor) {
    return Buffer.from(JSON.stringify(value)).toString('base64');
  },

  parseLiteral(ast): Cursor | null {
    if (ast.kind === Kind.STRING) {
      return JSON.parse(Buffer.from(ast.value, 'base64').toString('ascii'));
    }
    return null;
  },
});
```

**Argument in my resolver:**

```ts
@Arg('cursor', type => CursorScalar, { nullable: true })
cursor: Cursor
```

**Error:**

```
TypeError: Target is not a constructor
    at Object.convertToType (/app/node_modules/type-graphql/helpers/types.js:52:26)
    at Object.<anonymous> (/app/node_modules/type-graphql/resolvers/helpers.js:22:69)
    at Generator.next (<anonymous>)
    at /app/node_modules/type-graphql/resolvers/helpers.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/app/node_modules/type-graphql/resolvers/helpers.js:3:12)
    at Promise.all.params.sort.map (/app/node_modules/type-graphql/resolvers/helpers.js:17:88)
    at Array.map (<anonymous>)
    at Object.<anonymous> (/app/node_modules/type-graphql/resolvers/helpers.js:17:69)
    at Generator.next (<anonymous>)
    at /app/node_modules/type-graphql/resolvers/helpers.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (/app/node_modules/type-graphql/resolvers/helpers.js:3:12)
    at Object.getParams (/app/node_modules/type-graphql/resolvers/helpers.js:16:12)
    at /app/node_modules/type-graphql/resolvers/create.js:23:44
    at Generator.next (<anonymous>)
```

Am I doing something wrong @19majkel94 ?

**Comment by ashleyw** [comment]
There's a discrepancy between `GraphQLScalarType` used in my project and type-graphql.

```ts
  const schema = await TypeGraphQL.buildSchema({
    resolvers: [`${__dirname}/resolvers/**/*.ts`],
    scalarsMap: [{ type: Cursor, scalar: CursorScalar }],
  });
```

This results in a type warning: `Property 'toJSON' is missing in type 'GraphQLScalarType'`.

Which is strange, since my `package.json` defines `^0.13.2` (the same as type-graphql.)

**Comment by 19majkel94** [comment]
Try `npm dedupe` to flatten the tree and use the same module.

Maybe I should hide `graphql-js` and just export `GraphQLScalarType` and other from `type-graphql`.

**Comment by ashleyw** [comment]
I had to upgrade my `@types/graphql` package, but I'm still experiencing the issue. The `type instanceof GraphQLScalarType` check you mentioned resolves as true.

**Comment by 19majkel94** [comment]
Can you compile the app using `tsc` and see what TS reflection generates for this in your app:
```
@Arg('cursor', type => CursorScalar, { nullable: true })
cursor: Cursor
```
It should look something like this:
```ts
__decorate([
    src_1.Query(returns => [recipe_type_1.Recipe]),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], ExampleResolver.prototype, "recipes", null);
__decorate([
    src_1.Authorized() // only logged users can add new recipe
    ,
    src_1.Mutation(),
    __param(0, src_1.Arg("title")),
    __param(1, src_1.Arg("description", { nullable: true })),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, String]),
    __metadata("design:returntype", recipe_type_1.Recipe)
], ExampleResolver.prototype, "addRecipe", null);
__decorate([
    src_1.Authorized("ADMIN") // only admin can remove the published recipe
    ,
    src_1.Mutation(),
    __param(0, src_1.Arg("title")),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Boolean)
], ExampleResolver.prototype, "deleteRecipe", null);
ExampleResolver = __decorate([
    src_1.Resolver()
], ExampleResolver);
exports.ExampleResolver = ExampleResolver;
````

I wounder to what the `Cursor` type is reflected and why it can't be newable. I could add this to exclusion list.

**Comment by ashleyw** [comment]
```ts
__decorate([
    type_graphql_1.FieldResolver(),
    __param(0, type_graphql_1.Root()),
    __param(1, type_graphql_1.Arg('cursor', type => Cursor_1.CursorScalar, { nullable: true })),
    __param(2, type_graphql_1.Arg('limit', { nullable: true })),
    __param(3, type_graphql_1.Arg('direction', type => PaginationArgs_1.PaginationDirection, { nullable: true })),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [University_1.University, Object, Number, String]),
    __metadata("design:returntype", Promise)
], UniversityResolver.prototype, "technologies", null);
```
I'm fairly new to Typescript so I'm not sure exactly what I'm looking at. Would it be the `Object` in `design:paramtypes`?

**Comment by 19majkel94** [comment]
Object might be ignored but it works:
```
PS C:\WINDOWS\system32> node
> const o = new Object()
undefined
> o
{}
> o.test = "test"
'test'
> o
{ test: 'test' }
>
```

Unfortunately I don't know from the stacktrace on which argument it failed. Can you create a minimal reproducible example for that case? I would then be able to debug it and fix 😉

**Comment by ashleyw** [comment]
Here you go: https://github.com/ashleyw/type-graphql-issue-65-example :)

**Comment by 19majkel94** [comment]
Ok, now I see that. You scalar returns custom object, which goes through the if that checks the type of data only, not the target type:
```ts
if (simpleTypes.includes(data.constructor)) {
  return data;
}
```
Fixed in 4176ed2 🎉 
If you need transforming scalar object to class, use `scalarsMap` feature 😉
